src/errors.rs

```diff
@@ -0,0 +1,345 @@
+    Warning,
+}
+
+impl FromStr for RustcErrorKind {
+    type Err = ();
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let s = s.to_uppercase();
+        // Some RustcErrorKinds has this colon, so we need to split it
+        // See this for example:
+        // https://github.com/rust-lang/rust/blob/master/tests/ui/async-await/in-trait/fn-not-async-err.rs#L9
+        let part0: &str = s
+            .split(':')
+            .next()
+            .expect("split always returns at least one element");
+        match part0 {
+            "HELP" => Ok(RustcErrorKind::Help),
+            "ERROR" => Ok(RustcErrorKind::Error),
+            "NOTE" => Ok(RustcErrorKind::Note),
+            "SUGGESTION" => Ok(RustcErrorKind::Suggestion),
+            "WARN" | "WARNING" => Ok(RustcErrorKind::Warning),
+            _ => Err(()),
+        }
+    }
+}
+
+impl fmt::Display for RustcErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            RustcErrorKind::Help => write!(f, "help message"),
+            RustcErrorKind::Error => write!(f, "error"),
+            RustcErrorKind::Note => write!(f, "note"),
+            RustcErrorKind::Suggestion => write!(f, "suggestion"),
+            RustcErrorKind::Warning => write!(f, "warning"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Error {
+    pub line_num: usize,
+    /// We also need to take into account the relative line number.
+    /// if the error is on the previous line, then it's value is `1`.
+    /// if the error is on the same line, then it's value is `0`.
+    /// if the error is on the next line, then it's value is `-1`.
+    pub relative_line_num: i32,
+
+    /// What kind of message we expect (e.g., warning, error, suggestion).
+    /// `None` if not specified or unknown message kind.
+    pub kind: Option<RustcErrorKind>,
+    pub msg: String,
+    pub error_code: Option<String>,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let error_code = self.error_code.as_ref().map_or("", |code| &code[..]);
+        write!(
+            f,
+            "// {{ {} \"{}\" \"\" {{ target *-*-* }} {}}}",
+            match &self.kind {
+                Some(kind) => match kind {
+                    RustcErrorKind::Help => "help",
+                    RustcErrorKind::Error => "dg-error",
+                    RustcErrorKind::Note => "dg-note",
+                    RustcErrorKind::Suggestion => "suggestion",
+                    RustcErrorKind::Warning => "dg-warning",
+                },
+                None => "dg-error",
+            },
+            if !error_code.is_empty() {
+                format!(".{}.", error_code)
+            } else {
+                error_code.to_owned()
+            },
+            if self.relative_line_num == 0 {
+                "".to_owned()
+            } else {
+                format!(".{} ", self.relative_line_num)
+            },
+        )
```

Comment:
I think you should create a temporary variable here. Yes rust is expression oriented, but we're not in lisp. When it becomes hard to read we can simply put temporary variables.
```suggestion
        use RustcErrorKind::*;
        let error_type = match &self.kind {
            Some(Help) => "help",
            Some(Error) => "dg-error",
            Some(Note) => "dg-note",
            Some(Suggestion) => "suggestion",
            Some(Warning) => "dg-warning",
            None => "dg-error",
        };
        
        let error_code = if error_code.is_empty() {
        // Note I've changed the order: when possible we want to avoid negations and complex expressions
            error_code.to_owned()
        } else {
            format!(".{}.", error_code)
        };
        
        let rel_line_number = if self.relative_line_num == 0 {
            "".to_owned()
        } else {
            format!(".{} ", self.relative_line_num)
        };
            
        write!(f, "// {{ {error_type} \"{error_code}\" \"\" {{ target *-*-* }} {rel_line_number}}}")
```